tsdb/db.go

```diff
@@ -1516,8 +1516,29 @@ func (db *DB) Querier(_ context.Context, mint, maxt int64) (storage.Querier, err
 			blocks = append(blocks, b)
 		}
 	}
+	var headQuerier storage.Querier
 	if maxt >= db.head.MinTime() {
-		blocks = append(blocks, NewRangeHead(db.head, mint, maxt))
+		rh := NewRangeHead(db.head, mint, maxt)
+		var err error
+		headQuerier, err = NewBlockQuerier(rh, mint, maxt)
+		if err != nil {
+			return nil, errors.Wrapf(err, "open querier for head %s", rh)
+		}
+
+		shouldClose, getNew, newMint := db.head.IsQuerierValid(mint, maxt)
```

Comment:
I might be missing a locking layer, but it looks to me like a truncation could start just after this line has returned "all good" (i.e. `false, false, 0`), and then we would run into the same problem again. (In other words, is still a race, just that we reduced the time window in which the race could be triggered.)

Reply:
Good question. So that is taken care by first getting the querier above before calling this (as getting querier will register itself in the isolation state linked list). So if we got `false, false, 0`, that means truncation had not started yet. So even if truncation starts after this, it will wait at `WaitForPendingReadersInTimeRange` since querier was registered before this point.

Reply:
I will add a comment in the code regarding this.

Reply:
OK, that might be the "locking layer" I was missing. I will have another look and try to wrap my head around it.